invisible_cities/io/mcinfo_io.py

```
@@ -187,26 +280,93 @@ def copy_mc_info(h5in : tb.File, writer : Type[mc_info_writer], which_events : L
+            new_key = MCTableType.event_mapping
+            tbl_dict[new_key] = pd.DataFrame(which_events, columns=['event_id'])
+        writer(tbl_dict)
     except tb.exceptions.NoSuchNodeError:
-        raise tb.exceptions.NoSuchNodeError(f"No MC info in file {h5in}.")
+        raise tb.exceptions.NoSuchNodeError(f"No MC info in file {file_in}.")
     except IndexError:
-        raise IndexError(f"No event {n} in file {h5in}.")
+        raise IndexError(f"No events {which_events} in file {file_in}.")
     except NoParticleInfoInFile as err:
         print(f"Warning: {h5in}", err)
         pass
 
 
-def load_mchits(file_name: str,
-                event_range=(0, int(1e9))) -> Mapping[int, Sequence[MCHit]]:
+def is_oldformat_file(file_name : str) -> bool:
+    with tb.open_file(file_name) as h5in:
+        return hasattr(h5in.root, 'MC/extents')
+
+
+def get_mc_tbl_list(file_name: str) -> List[MCTableType]:
+    """
+    Returns a list of the tables in
+    the MC group of a given file
+    """
+    with tb.open_file(file_name, 'r') as h5in:
+        mc_group = h5in.root.MC
+        return [MCTableType[tbl.name] for tbl in mc_group]
+
+
+def get_event_numbers_in_file(file_name: str) -> np.ndarray:
+    with tb.open_file(file_name, 'r') as h5in:
+        if is_oldformat_file(file_name):
+            evt_list = h5in.root.MC.extents.cols.evt_number[:]
+        else:
+            evt_list = _get_list_of_events_new(h5in)
+        return evt_list
+
+
+def _get_list_of_events_new(h5in : tb.file.File) -> np.ndarray:
+    mc_tbls  = ['hits', 'particles', 'sns_response']
+    def try_unique_evt_itr(group, itr):
+        for elem in itr:
+            try:
+                yield np.unique(getattr(group, elem).cols.event_id)
+            except tb.exceptions.NoSuchNodeError:
+                pass
+
+    evt_list = list(try_unique_evt_itr(h5in.root.MC, mc_tbls))
+    if len(evt_list) == 0:
+        raise tb.exceptions.NoSuchNodeError
```

this is a bit confusing, I am not sure that this function should raise an error if there are no tables with events or simply return an empty list?
In any case it is more clear to first get the name of the present tables and iterate over them extending the unique numpy array with `setdiff1d`or so

---

The exception was to maintain the expected try except structure approved for PR #702 but if you prefer I can look for ways to improve it?

We don't have access to the file name here to use the table list function and I'd prefer not to look at all tables in any case since many don't have an event_id column.

As to the extending, it seems like it'd need more gymnastics to do it that way since we'd have to initialise, right?